### PR TITLE
Cancel spot requests on timeout + add check for oversubscribed [ #32454645494108 ]

### PR DESCRIPTION
**Pull request by filipenf**

### CODE REVIEW

File: tools/spark-ec2/spark_ec2.py [hunk 1]
```diff
@@ -545,7 +543,11 @@ def launch_cluster(conn, opts, cluster_name):
                 else:
                     print "%d of %d slaves granted, waiting longer" % (
                         len(active_instance_ids), opts.slaves)
+
+                if (datetime.now() - start_time).seconds > (45*60):
```
Comment (douglaz) [inline tools/spark-ec2/spark_ec2.py]: Just make this a parameter.

File: tools/spark-ec2/spark_ec2.py [hunk 2]
```diff
@@ -523,18 +526,16 @@ def launch_cluster(conn, opts, cluster_name):
...
-                id_to_req = {}
-                for r in reqs:
-                    id_to_req[r.id] = r
-                active_instance_ids = []
-                for i in my_req_ids:
-                    if i in id_to_req and id_to_req[i].state == "active":
-                        active_instance_ids.append(id_to_req[i].instance_id)
+                reqs = conn.get_all_spot_instance_requests(my_req_ids)
+                active_instance_ids = filter(lambda req: req.state == "active", reqs)
+                oversubscribed = len(filter(lambda req: req.status.code == "capacity-oversubscribed", reqs))
```
Comment (douglaz) [inline tools/spark-ec2/spark_ec2.py]: Could you test also for: 'capacity-not-available' ?

Comment (douglaz) [inline tools/spark-ec2/spark_ec2.py]: Also: 'price-too-low'

Comment (filipenf) [inline tools/spark-ec2/spark_ec2.py]: Should we stop waiting for both cases?

Comment (douglaz) [inline tools/spark-ec2/spark_ec2.py]: Yes, abort.

**Comment by filipenf** [comment]
@douglaz Done. As a plus, printing the status message will give some additional information about the problem like: "Your Spot request price of 0.1 is lower than the minimum required Spot request fulfillment price of 2.1" :dart:

**Comment by douglaz** [comment]
LGTM!

**Pull request merged**